xdsl/interpreter.py

```diff
@@ -0,0 +1,273 @@
+    ```
+    """
+
+    @classmethod
+    def impls(
+        cls
+    ) -> Iterable[tuple[type[Operation], OpImpl[InterpreterFunctions,
+                                                Operation]]]:
+        ...
+
+
+_FT = TypeVar('_FT', bound=InterpreterFunctions)
+
+_IMPL_OP_TYPE = '__impl_op_type'
+_IMPL_DICT = '__impl_dict'
+
+P = ParamSpec('P')
+
+
+def impl(
+    op_type: type[OperationInvT]
+) -> Callable[[OpImpl[_FT, OperationInvT]], OpImpl[_FT, OperationInvT]]:
+    "See `InterpreterFunctions`"
+
+    def annot(func: OpImpl[_FT, OperationInvT]) -> OpImpl[_FT, OperationInvT]:
+        setattr(func, _IMPL_OP_TYPE, op_type)
+        return func
+
+    return annot
+
+
+def register_impls(ft: type[_FT]) -> type[_FT]:
+    "See `InterpreterFunctions`"
+    impl_dict: _ImplDict = {}
+    for cls in ft.mro():
+        # Iterate from subclass through superclasses
+        # Assign definitions, unless they've been redefined in a subclass
+        for val in cls.__dict__.values():
+            if _IMPL_OP_TYPE in val.__dir__():
+                # This is an annotated function
+                op_type = getattr(val, _IMPL_OP_TYPE)
+                if op_type not in impl_dict:
+                    # subclass overrides superclass definition
+                    impl_dict[op_type] = val  # type: ignore
+    setattr(ft, _IMPL_DICT, impl_dict)
+
+    @classmethod
+    def impls(
+        cls: type[_FT]
+    ) -> Iterable[tuple[type[Operation], OpImpl[InterpreterFunctions,
+                                                Operation]]]:
+        return impl_dict.items()
+
+    setattr(ft, 'impls', impls)
+    return ft
+
+
+@dataclass
+class _InterpreterFunctionImpls:
+    """
+    Used to combine multiple function implementations. The operation
+    implementations need to be passed the instance of the Functions class,
+    so we keep a `(Functions, OpImpl)` tuple for every Operation type.
+    """
+
+    _impl_dict: dict[type[Operation],
+                     tuple[InterpreterFunctions,
+                           OpImpl[InterpreterFunctions,
+                                  Operation]]] = field(default_factory=dict)
+
+    def register_from(self, ft: InterpreterFunctions, /, override: bool):
+        for op_type, impl in ft.impls():
+            if op_type in self._impl_dict and not override:
+                raise ValueError(
+                    "Attempting to register implementation for op of type "
+                    f"{op_type}, but type already registered")
+
+            self._impl_dict[op_type] = (ft, impl)
+
+    def run(self, interpreter: Interpreter, op: Operation,
+            args: tuple[Any, ...]) -> tuple[Any, ...]:
+        if type(op) not in self._impl_dict:
+            raise InterpretationError(
+                f'Could not find interpretation function for op {op.name}')
+        ft, impl = self._impl_dict[type(op)]
+        return impl(ft, interpreter, op, args)
+
+
+@dataclass
+class InterpreterContext:
```

Comment:
I would maybe rename that to `InterpreterEnvironment`?
But I don't feel strongly about it.